### ISSUE TITLE
feat: enable ggml plugin support on Amazon Linux 2

### ISSUE THREAD

**Issue by hydai** [WasmEdge/WasmEdge#3034]
### Summary

Some users work on AWS Deep Learning AMI based on Amazon Linux 2. When they tried our wasmedge+ggml examples, they failed due to several incompatible issues.
To avoid these kinds of issues happening again, we would like to have a new release and building assets for the Amazon Linux 2.

### Details

Major:

- [ ] Build WasmEdge on Amazon Linux 2
- [ ] Build ggml plugin with CUDA enabled option on Amazon Linux 2

Minor:

- [ ] Support the remaining plugins

### Appendix

- [Amazon Linux docker images](https://hub.docker.com/_/amazonlinux/)

**Comment by richzw** [comment]
@hydai After some investigation, I found one issue with Amazon Linux 2 is that the version of `glibc` is `2.26`

```
> ldd --version
ldd (GNU libc) 2.26
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

However, the version of `glibc` required by `libwasmedgePluginWasiNN` is `2.29`

```
> wasmedge --dir .:. --nn-preload default:GGML:AUTO:llama-2-7b-chat.Q5_K_M.gguf llama-chat.wasm
[2023-11-15 12:56:50.309] [error] loading failed: invalid path, Code: 0x20
[2023-11-15 12:56:50.309] [error]     load library failed:/lib64/libm.so.6: version `GLIBC_2.29' not found (required by /home/ec2-user/.wasmedge/lib/../plugin/libwasmedgePluginWasiNN.so)
```

The os version is 

```
> cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
SUPPORT_END="2025-06-30"
```

Please correct me if something missing or wrong.

**Comment by hydai** [comment]
> @hydai After some investigation, I found one issue with Amazon Linux 2 is that the version of `glibc` is `2.26`
> 

Yes. So we have to rebuild and add CI for amazon Linux 2, it can make sure the `glibc` is the same as the AMI one.

**Comment by richzw** [comment]
> > @hydai After some investigation, I found one issue with Amazon Linux 2 is that the version of `glibc` is `2.26`
> 
> Yes. So we have to rebuild and add CI for amazon Linux 2, it can make sure the `glibc` is the same as the AMI one.

Got it. Thank you very much for your reply.

**Comment by 0x-avi** [comment]
@hydai may i work on this?
if yes i just wanted to clear the deliverables :
- a docker image which builds ggml with cuda and also wasmedge build
- an asset (gzip) file for the Amazon Linux2

questions:
- do we also need to build for amazon linux 2023 and amazon linux AMI?
- is the process of doing everything the same #3707 ?

thanks

**Comment by hydai** [comment]
> may i work on this?

Yes

> * a docker image which builds ggml with cuda and also wasmedge build

There should be a Amazon Linux 2 image on the dockerhub. The target is to ensure that WasmEdge + ggml plugin can be built and executed on Amazon Linux 2

> * an asset (gzip) file for the Amazon Linux2

I don't get the point.

> * do we also need to build for amazon linux 2023 and amazon linux AMI?

Amazon linux AMI is EOL. We have zero interest in it.
It should be easier to build for Amazon Linux 2023 because it offers a more modern software stack.

> * is the process of doing everything the same [[CI][Docker] Add aarch64 for ubuntu 20.04 #3707](https://github.com/WasmEdge/WasmEdge/pull/3707) ?

Yes.

**Comment by hydai** [comment]
Closing this due to Amazon Linux 2 reaching EOL in five months.
